### ISSUE TITLE
Accessible schema properties

### ISSUE THREAD

**Issue by brandur** [xeipuuv/gojsonschema#15]
Nearly all fields in gojsonschema structs are invisible outside the package. This works pretty well as long as its mainly used to call `Validate`, but it might be useful to be able to re-use gojsonschema to explore a document outside of this context. For example, if I've built a JSON Hyper Schema and would like to enumerate every link that it provides.

Have you put any thought into opening up these interfaces for re-use?

**Comment by abecciu** [comment]
+1 this would be super useful!

**Comment by shutej** [comment]
I agree.  I'm currently looking at:

https://github.com/interagent/schematic

This seems to be developed by the Heroku folks for use on the Heroku-created Hyper Schema.  They've re-implemented some of this package, but because Heroku's schema is now self-referential, their own code generator seems to ... not work.

I would think the design of this might go like so:

- [ ] Make a Visitor interface and a Visit function on a parsed schema.
- [ ] Make a ValidationVisitor struct that implements the visitor interface and tracks errors and validity internally.
- [ ] Make the old validation functions into helpers that instantiate a ValidationVisitor and then call Visit on the schema.

This would then let other users instantiate their own visitors.

**Comment by shutej** [comment]
@xeipuuv would you be open to a pull request that moves the design in this direction?

**Comment by xeipuuv** [comment]
Yes i am open to pull requests

**Comment by brandur** [comment]
> This seems to be developed by the Heroku folks for use on the Heroku-created Hyper Schema. They've re-implemented some of this package, but because Heroku's schema is now self-referential, their own code generator seems to ... not work.

Yeah, this one is our bad :/ Essentially that project was booted off, but didn't really have a story around longer term maintenance and eventually it stopped working and wasn't fixed. I'm somewhat hopeful that some of the internal reorganization that we're doing will eventually put this in a position where it has some more dedicated maintainers, but that's a few months out at minimum unfortunately.

For what it's worth, I really like the proposed design above. Projects like Schematic shouldn't even think about trying to parse and interpret a JSON Schema for themselves and should instead rely on more complete and better vetted libraries like this one; having a generic visitor pattern would allow us to get there.
